File: packages/runar-compiler/src/__tests__/readonly-property-assignment.test.ts
// ---------------------------------------------------------------------------
// Audit C2 — `readonly` property assignment must be rejected outside the
// constructor.
//
// spec/semantics.md:247   <this.p = e, env, sigma> ==> ERROR: cannot assign to
//                         readonly property
// spec/grammar.md:94      readonly properties "cannot be reassigned"
//
// Before this rule existed, a contract that reassigned its readonly owner
// before checking it compiled to `76a97ca9788777` — hash160(pk) compared
// against hash160(pk), true for ANY pubkey, i.e. anyone can spend.
//
// The constructor MUST still be allowed to assign readonly properties.
//
// Mirrored in the peer tiers:
//   compilers/go/frontend/validator_test.go
//   compilers/rust/src/frontend/validator.rs      (#[cfg(test)] mod tests)
//   compilers/python/tests/test_validator.py
//   compilers/zig/src/passes/validate.zig         (inline test blocks)
//   compilers/ruby/test/test_validator.rb
//   compilers/java/src/test/java/runar/compiler/passes/ValidateTest.java
// ---------------------------------------------------------------------------

import { describe, it, expect } from 'vitest';
import { parse } from '../passes/01-parse.js';
import { validate } from '../passes/02-validate.js';
import type { ValidationResult } from '../passes/02-validate.js';
import type { ContractNode } from '../ir/index.js';

function parseContract(source: string, fileName?: string): ContractNode {
  const result = parse(source, fileName);
  if (!result.contract) {
    throw new Error(`Parse failed: ${result.errors.map(e => e.message).join(', ')}`);
  }
  return result.contract;
}

function validateSource(source: string, fileName?: string): ValidationResult {
  return validate(parseContract(source, fileName));
}

function hasError(result: ValidationResult, substring: string): boolean {
  return result.errors.some(e => e.message.includes(substring));
}

/** The common cross-tier diagnostic substring. */
const READONLY_WRITE = 'assign to readonly property';

describe('C2: readonly property assignment', () => {
  it('rejects the owner-hijack contract that reassigns its readonly owner', () => {
    // Compiles to 76a97ca9788777 without the rule: hash160(pk) == hash160(pk).
    const source = `
class Hijack extends SmartContract {
  readonly ownerHash: Addr;

  constructor(ownerHash: Addr) {
    super(ownerHash);
    this.ownerHash = ownerHash;
  }

  public unlock(attackerPk: PubKey) {
    this.ownerHash = hash160(attackerPk);
    assert(hash160(attackerPk) == this.ownerHash);
  }
}
`;
    const result = validateSource(source);
    expect(hasError(result, READONLY_WRITE)).toBe(true);
  });

  it('reports the offending property name', () => {
    const source = `
class Hijack extends SmartContract {
  readonly ownerHash: Addr;

  constructor(ownerHash: Addr) {
    super(ownerHash);
    this.ownerHash = ownerHash;
  }

  public unlock(attackerPk: PubKey) {
    this.ownerHash = hash160(attackerPk);
    assert(true);
  }
}
`;
    const result = validateSource(source);
    expect(result.errors.some(e => e.message.includes("'ownerHash'"))).toBe(true);
  });

  it('rejects a readonly write in a StatefulSmartContract method', () => {
    const source = `
class Vault extends StatefulSmartContract {
  readonly owner: Addr;
  count: bigint;

  constructor(owner: Addr, count: bigint) {
    super(owner, count);
    this.owner = owner;
    this.count = count;
  }

  public bump(newOwner: Addr) {
    this.owner = newOwner;
    this.count = this.count + 1n;
  }
}
`;
    const result = validateSource(source);
    expect(hasError(result, READONLY_WRITE)).toBe(true);
  });

  it('rejects a readonly write nested inside an if branch', () => {
    const source = `
class Nested extends StatefulSmartContract {
  readonly owner: Addr;
  count: bigint;

  constructor(owner: Addr, count: bigint) {
    super(owner, count);
    this.owner = owner;
    this.count = count;
  }

  public bump(newOwner: Addr, flag: boolean) {
    if (flag) {
      this.owner = newOwner;
    } else {
      this.count = this.count + 1n;
    }
  }
}
`;
    const result = validateSource(source);
    expect(hasError(result, READONLY_WRITE)).toBe(true);
  });

  it('rejects a readonly write in a private helper method', () => {
    const source = `
class Helper extends StatefulSmartContract {
  readonly owner: Addr;
  count: bigint;

  constructor(owner: Addr, count: bigint) {
    super(owner, count);
    this.owner = owner;
    this.count = count;
  }

  private steal(newOwner: Addr): void {
    this.owner = newOwner;
  }

  public bump(newOwner: Addr) {
    this.steal(newOwner);
    this.count = this.count + 1n;
  }
}
`;
    const result = validateSource(source);
    expect(hasError(result, READONLY_WRITE)).toBe(true);
  });

  it('rejects increment/decrement of a readonly property', () => {
    const source = `
class Bump extends StatefulSmartContract {
  readonly limit: bigint;
  count: bigint;

  constructor(limit: bigint, count: bigint) {
    super(limit, count);
    this.limit = limit;
    this.count = count;
  }

  public go() {
    this.limit++;
    this.count = this.count + 1n;
  }
}
`;
    const result = validateSource(source);
    expect(hasError(result, READONLY_WRITE)).toBe(true);
  });

  it('rejects the Python-surface hijack contract too', () => {
    const source = `
from runar import SmartContract, assert_, hash160, Readonly, ByteString, PubKey

class Hijack(SmartContract):
    owner_hash: Readonly[ByteString]

    def __init__(self, owner_hash: ByteString):
        super().__init__(owner_hash)
        self.owner_hash = owner_hash

    @public
    def unlock(self, attacker_pk: PubKey):
        self.owner_hash = hash160(attacker_pk)
        assert_(hash160(attacker_pk) == self.owner_hash)
`;
    const result = validateSource(source, 'Hijack.runar.py');
    expect(hasError(result, READONLY_WRITE)).toBe(true);
  });

  // -------------------------------------------------------------------------
  // The constructor must keep working — every contract assigns its readonly
  // properties there.
  // -------------------------------------------------------------------------

  it('accepts readonly assignment in the constructor', () => {
    const source = `
class P2PKH extends SmartContract {
  readonly pubKeyHash: Addr;

  constructor(pubKeyHash: Addr) {
    super(pubKeyHash);
    this.pubKeyHash = pubKeyHash;
  }

  public unlock(sig: Sig, pk: PubKey) {
    assert(hash160(pk) == this.pubKeyHash);
    assert(checkSig(sig, pk));
  }
}
`;
    const result = validateSource(source);
    expect(result.errors).toEqual([]);
  });

  it('accepts mutable state mutation in a StatefulSmartContract method', () => {
    const source = `
class Counter extends StatefulSmartContract {
  readonly owner: Addr;
  count: bigint;

  constructor(owner: Addr, count: bigint) {
    super(owner, count);
    this.owner = owner;
    this.count = count;
  }

  public increment() {
    this.count = this.count + 1n;
  }
}
`;
    const result = validateSource(source);
    expect(result.errors).toEqual([]);
  });

  it('accepts reading a readonly property in a method', () => {
    const source = `
class Reader extends StatefulSmartContract {
  readonly limit: bigint;
  count: bigint;

  constructor(limit: bigint, count: bigint) {
    super(limit, count);
    this.limit = limit;
    this.count = count;
  }

  public increment() {
    assert(this.count < this.limit);
    this.count = this.count + 1n;
  }
}
`;
    const result = validateSource(source);
    expect(result.errors).toEqual([]);
  });

  it('accepts a local variable that shadows a readonly property name', () => {
    const source = `
class Shadow extends StatefulSmartContract {
  readonly limit: bigint;
  count: bigint;

  constructor(limit: bigint, count: bigint) {
    super(limit, count);
    this.limit = limit;
    this.count = count;
  }

  public increment() {
    let limit: bigint = 5n;
    limit = 6n;
    assert(this.count < limit);
    this.count = this.count + 1n;
  }
}
`;
    const result = validateSource(source);
    expect(result.errors).toEqual([]);
  });
});

File: compilers/java/src/test/java/runar/compiler/passes/ReadonlyPropertyAssignmentTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;
import runar.compiler.frontend.ParserDispatch;
import runar.compiler.ir.ast.ContractNode;

/**
 * Audit C2 — {@code readonly} property assignment must be rejected outside the
 * constructor.
 *
 * <p>{@code spec/semantics.md:247}:
 * {@code <this.p = e, env, sigma> ==> ERROR: cannot assign to readonly property}
 *
 * <p>Without the rule a contract that reassigns its readonly owner before
 * checking it compiles to {@code 76a97ca9788777} — hash160(pk) compared against
 * hash160(pk), true for ANY pubkey, i.e. anyone can spend.
 *
 * <p>The constructor MUST still be allowed to assign readonly properties.
 *
 * <p>Mirrors {@code packages/runar-compiler/src/__tests__/readonly-property-assignment.test.ts}.
 */
class ReadonlyPropertyAssignmentTest {

    /** The cross-tier diagnostic substring. */
    private static final String READONLY_WRITE = "assign to readonly property";

    private static Validate.Result validateSource(String source, String fileName) throws Exception {
        ContractNode c = ParserDispatch.parse(source, fileName);
        return Validate.runCollecting(c);
    }

    private static void assertReadonlyWriteError(Validate.Result r) {
        assertTrue(
            r.errors().stream().anyMatch(m -> m.contains(READONLY_WRITE)),
            "expected a readonly-write error, got: " + r.errors()
        );
    }

    @Test
    void rejectsOwnerHijackContract() throws Exception {
        String src = """
            import { SmartContract, Addr, PubKey } from 'runar-lang';

            class Hijack extends SmartContract {
              readonly ownerHash: Addr;

              constructor(ownerHash: Addr) {
                super(ownerHash);
                this.ownerHash = ownerHash;
              }

              public unlock(attackerPk: PubKey) {
                this.ownerHash = hash160(attackerPk);
                assert(hash160(attackerPk) === this.ownerHash);
              }
            }
            """;
        Validate.Result r = validateSource(src, "Hijack.runar.ts");
        assertReadonlyWriteError(r);
        assertTrue(
            r.errors().stream().anyMatch(m -> m.contains("'ownerHash'")),
            "expected the offending property name, got: " + r.errors()
        );
    }

    @Test
    void rejectsReadonlyWriteInStatefulMethod() throws Exception {
        String src = """
            import { StatefulSmartContract, Addr } from 'runar-lang';

            class Vault extends StatefulSmartContract {
              readonly owner: Addr;
              count: bigint;

              constructor(owner: Addr, count: bigint) {
                super(owner, count);
                this.owner = owner;
                this.count = count;
              }

              public bump(newOwner: Addr) {
                this.owner = newOwner;
                this.count = this.count + 1n;
              }
            }
            """;
        assertReadonlyWriteError(validateSource(src, "Vault.runar.ts"));
    }

    @Test
    void rejectsReadonlyWriteNestedInIf() throws Exception {
        String src = """
            import { StatefulSmartContract, Addr } from 'runar-lang';

            class Nested extends StatefulSmartContract {
              readonly owner: Addr;
              count: bigint;

              constructor(owner: Addr, count: bigint) {
                super(owner, count);
                this.owner = owner;
                this.count = count;
              }

              public bump(newOwner: Addr, flag: boolean) {
                if (flag) {
                  this.owner = newOwner;
                } else {
                  this.count = this.count + 1n;
                }
              }
            }
            """;
        assertReadonlyWriteError(validateSource(src, "Nested.runar.ts"));
    }

    @Test
    void rejectsReadonlyWriteInPrivateHelper() throws Exception {
        String src = """
            import { StatefulSmartContract, Addr } from 'runar-lang';

            class Helper extends StatefulSmartContract {
              readonly owner: Addr;
              count: bigint;

              constructor(owner: Addr, count: bigint) {
                super(owner, count);
                this.owner = owner;
                this.count = count;
              }

              private steal(newOwner: Addr): void {
                this.owner = newOwner;
              }

              public bump(newOwner: Addr) {
                this.steal(newOwner);
                this.count = this.count + 1n;
              }
            }
            """;
        assertReadonlyWriteError(validateSource(src, "Helper.runar.ts"));
    }

    @Test
    void rejectsIncrementOfReadonlyProperty() throws Exception {
        String src = """
            import { StatefulSmartContract } from 'runar-lang';

            class Bump extends StatefulSmartContract {
              readonly limit: bigint;
              count: bigint;

              constructor(limit: bigint, count: bigint) {
                super(limit, count);
                this.limit = limit;
                this.count = count;
              }

              public go() {
                this.limit++;
                this.count = this.count + 1n;
              }
            }
            """;
        assertReadonlyWriteError(validateSource(src, "Bump.runar.ts"));
    }

    @Test
    void rejectsPythonSurfaceHijackContract() throws Exception {
        String src = """
            from runar import SmartContract, assert_, hash160, Readonly, ByteString, PubKey

            class Hijack(SmartContract):
                owner_hash: Readonly[ByteString]

                def __init__(self, owner_hash: ByteString):
                    super().__init__(owner_hash)
                    self.owner_hash = owner_hash

                @public
                def unlock(self, attacker_pk: PubKey):
                    self.owner_hash = hash160(attacker_pk)
                    assert_(hash160(attacker_pk) == self.owner_hash)
            """;
        assertReadonlyWriteError(validateSource(src, "Hijack.runar.py"));
    }

    @Test
    void rejectsJavaSurfaceHijackContract() throws Exception {
        String src = """
            class Hijack extends SmartContract {
                @Readonly Addr ownerHash;

                Hijack(Addr ownerHash) {
                    super(ownerHash);
                    this.ownerHash = ownerHash;
                }

                @Public
                void unlock(PubKey attackerPk) {
                    this.ownerHash = hash160(attackerPk);
                    assertThat(hash160(attackerPk).equals(ownerHash));
                }
            }
            """;
        assertReadonlyWriteError(validateSource(src, "Hijack.runar.java"));
    }

    // -------------------------------------------------------------------------
    // The constructor must keep working — every contract assigns its readonly
    // properties there.
    // -------------------------------------------------------------------------

    @Test
    void acceptsReadonlyAssignmentInConstructor() throws Exception {
        String src = """
            import { SmartContract, Addr, PubKey, Sig } from 'runar-lang';

            class P2PKH extends SmartContract {
              readonly pubKeyHash: Addr;

              constructor(pubKeyHash: Addr) {
                super(pubKeyHash);
                this.pubKeyHash = pubKeyHash;
              }

              public unlock(sig: Sig, pubKey: PubKey) {
                assert(hash160(pubKey) === this.pubKeyHash);
                assert(checkSig(sig, pubKey));
              }
            }
            """;
        Validate.Result r = validateSource(src, "P2PKH.runar.ts");
        assertTrue(r.errors().isEmpty(), "expected no errors, got: " + r.errors());
    }

    @Test
    void acceptsMutableStateMutation() throws Exception {
        String src = """
            import { StatefulSmartContract, Addr } from 'runar-lang';

            class Counter extends StatefulSmartContract {
              readonly owner: Addr;
              count: bigint;

              constructor(owner: Addr, count: bigint) {
                super(owner, count);
                this.owner = owner;
                this.count = count;
              }

              public increment() {
                this.count = this.count + 1n;
              }
            }
            """;
        Validate.Result r = validateSource(src, "Counter.runar.ts");
        assertTrue(r.errors().isEmpty(), "expected no errors, got: " + r.errors());
    }

    @Test
    void acceptsLocalShadowingAReadonlyPropertyName() throws Exception {
        String src = """
            import { StatefulSmartContract } from 'runar-lang';

            class Shadow extends StatefulSmartContract {
              readonly limit: bigint;
              count: bigint;

              constructor(limit: bigint, count: bigint) {
                super(limit, count);
                this.limit = limit;
                this.count = count;
              }

              public increment() {
                let limit: bigint = 5n;
                limit = 6n;
                assert(this.count < limit);
                this.count = this.count + 1n;
              }
            }
            """;
        Validate.Result r = validateSource(src, "Shadow.runar.ts");
        assertTrue(r.errors().isEmpty(), "expected no errors, got: " + r.errors());
    }
}
